File: apps/api/app/fixtures/quality.py
from app.core.clock import Clock, parse_rfc3339
from app.schemas import AvailabilityStatus, DataQuality, FreshnessLevel

MOCK_SOURCE = "mock.fixtures.v1"
MOCK_NOTE = "Fixture fictive. Ne pas interpréter comme une donnée sportive réelle."


def quality(
    clock: Clock,
    *,
    availability: AvailabilityStatus = "available",
    source: str | None = MOCK_SOURCE,
    observed_at: str | None = "now",
    freshness: FreshnessLevel | None = "fresh",
    note: str | None = MOCK_NOTE,
) -> DataQuality:
    observed = None
    if availability == "unavailable":
        source = None
        freshness = None
        if note == MOCK_NOTE:
            note = "Donnée absente du jeu mock. Ce n'est pas une valeur nulle."
    elif observed_at == "now":
        observed = clock.now()
    elif observed_at is not None:
        observed = parse_rfc3339(observed_at)

    if availability == "stale" and freshness is None:
        freshness = "stale"

    return DataQuality(
        availability=availability,
        source=source,
        observed_at=observed,
        freshness=freshness,
        note=note,
    )


def unavailable(clock: Clock) -> DataQuality:
    return quality(clock, availability="unavailable")

File: apps/web/src/data/mock/quality.ts
import type { AvailabilityStatus, DataQuality, FreshnessLevel } from "@/types/api";
import { MOCK_NOW_ISO } from "@/data/mock/clock";

export function quality(options: {
  availability?: AvailabilityStatus;
  source?: string | null;
  observed_at?: string | null;
  freshness?: FreshnessLevel | null;
  note?: string | null;
} = {}): DataQuality {
  const availability = options.availability ?? "available";

  return {
    availability,
    source: options.source ?? (availability === "unavailable" ? null : "mock.fixtures"),
    observed_at:
      options.observed_at ?? (availability === "unavailable" ? null : MOCK_NOW_ISO),
    freshness:
      options.freshness ??
      (availability === "stale" ? "stale" : availability === "unavailable" ? null : "fresh"),
    note: options.note ?? null,
  };
}

export const unavailable = quality({
  availability: "unavailable",
  source: null,
  observed_at: null,
  freshness: null,
  note: "Donnée absente du jeu mock. Ce n'est pas une valeur nulle.",
});
